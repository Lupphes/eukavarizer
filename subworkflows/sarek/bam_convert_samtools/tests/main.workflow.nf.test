nextflow_workflow {

    name "Test Workflow BAM_CONVERT_SAMTOOLS"
    script "../main.nf"
    workflow "BAM_CONVERT_SAMTOOLS"
    config "subworkflows/sarek/bam_convert_samtools/tests/nextflow.config"
    tag "subworkflows"
    tag "subworkflows_sarek"
    tag "bam_convert_samtools"

    /*
     * ============================================================================
     * PAIRED-END TESTS
     * ============================================================================
     */

    /*
     * Test scenario 1: BAM to FASTQ conversion (paired-end) - Non-interleaved
     * This test validates BAM file conversion to paired-end FASTQ files
     * Uses test.paired_end.sorted.bam from nf-core test datasets
     * Expected behavior: Extract mapped and unmapped reads, merge unmapped reads, and output paired FASTQ files
     * Data structure: COLLATE_FASTQ outputs [R1, R2], CAT_FASTQ receives [mapped_R1, mapped_R2, unmapped_R1, unmapped_R2]
     */
    test("BAM to FASTQ - paired end - stub") {
        options '-stub'

        when {
            workflow {
                """
                // BAM file with paired-end reads (index will be generated by workflow)
                // Using tuple structure: [meta, bam_file]
                input[0] = channel.of([
                    [id: 'test', single_end: false],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                ])
                // Reference genome (required for samtools operations)
                input[1] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                // Reference genome index
                input[2] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ])
                // Interleaved flag (false = separate R1/R2 files)
                input[3] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.reads != null }
            )
        }
    }

    /*
     * Test scenario 2: BAM to FASTQ conversion with interleaved output
     * This test validates BAM file conversion to interleaved FASTQ format
     * Expected behavior: Extract and merge reads into interleaved FASTQ format (R1/R2 in same file)
     */
    test("BAM to FASTQ - interleaved - stub") {
        options '-stub'

        when {
            workflow {
                """
                input[0] = channel.of([
                    [id: 'test_interleaved', single_end: false],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                ])
                input[1] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ])
                input[3] = true  // interleaved = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.reads != null }
            )
        }
    }

    /*
     * Test scenario 3: CRAM to FASTQ conversion
     * This test validates CRAM file conversion to FASTQ files with reference genome
     * CRAM format requires reference genome for decompression
     * Expected behavior: Extract reads from CRAM using reference, output FASTQ files
     */
    test("CRAM to FASTQ - with reference - stub") {
        options '-stub'

        when {
            workflow {
                """
                input[0] = channel.of([
                    [id: 'test_cram', single_end: false],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram', checkIfExists: true)
                ])
                input[1] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true)
                ])
                input[3] = false  // interleaved = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.reads != null }
            )
        }
    }

    /*
     * ============================================================================
     * SINGLE-END TESTS
     * ============================================================================
     */

    /*
     * Test scenario 4: BAM to FASTQ conversion (single-end)
     * This test validates BAM file conversion for single-end sequencing data
     * Uses test.single_end.bam from nf-core test datasets
     * Expected behavior: Extract mapped and unmapped reads, merge them into single output
     * Data structure: COLLATE_FASTQ outputs single file, CAT_FASTQ receives [mapped, unmapped] flat list
     */
    test("BAM to FASTQ - single end - stub") {
        options '-stub'

        when {
            workflow {
                """
                input[0] = channel.of([
                    [id: 'test_se', single_end: true],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.bam', checkIfExists: true)
                ])
                input[1] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                ])
                input[3] = false  // interleaved = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.reads != null }
            )
        }
    }

    /*
     * Test scenario 5: BAM to FASTQ conversion (single-end) - Multiple samples
     * This test validates single-end BAM conversion with multiple samples
     * Tests that the workflow properly handles multiple single-end inputs
     * Expected behavior: Process both samples independently and output separate FASTQ files
     */
    test("BAM to FASTQ - single end multiple samples - stub") {
        options '-stub'

        when {
            workflow {
                """
                input[0] = channel.of(
                    [
                        [id: 'test_se1', single_end: true],
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.bam', checkIfExists: true)
                    ],
                    [
                        [id: 'test_se2', single_end: true],
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.bam', checkIfExists: true)
                    ]
                )
                input[1] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[2] = channel.value([
                    [id: 'genome'],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                ])
                input[3] = false  // interleaved = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.reads != null }
            )
        }
    }
}
