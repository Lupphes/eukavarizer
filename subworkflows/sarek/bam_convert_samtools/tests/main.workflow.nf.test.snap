{
    "BAM to FASTQ - paired end - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    "versions.yml:md5,046b30344f9a090d641ff7f591b0ee50",
                    "versions.yml:md5,3ec86bd3456e93b6fba8f3ee9c1d127e",
                    "versions.yml:md5,4db4e520adf0ec14c0428d440d72d0b3",
                    "versions.yml:md5,78e78e1ece80eef67ae4fbca2fdfbb70",
                    "versions.yml:md5,8a3063fe8a5ecf5275b33d92003b5508",
                    "versions.yml:md5,b97473987bd71717f820aa380ed74e70",
                    "versions.yml:md5,ca2df4d2c212dad9ab2a8d50c45b35c8"
                ],
                "reads": [
                    
                ],
                "versions": [
                    "versions.yml:md5,046b30344f9a090d641ff7f591b0ee50",
                    "versions.yml:md5,3ec86bd3456e93b6fba8f3ee9c1d127e",
                    "versions.yml:md5,4db4e520adf0ec14c0428d440d72d0b3",
                    "versions.yml:md5,78e78e1ece80eef67ae4fbca2fdfbb70",
                    "versions.yml:md5,8a3063fe8a5ecf5275b33d92003b5508",
                    "versions.yml:md5,b97473987bd71717f820aa380ed74e70",
                    "versions.yml:md5,ca2df4d2c212dad9ab2a8d50c45b35c8"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-22T01:43:55.749869932"
    }
}